### PR TITLE
[4.0] Field module order moved to custom elements

### PR DESCRIPTION
**Pull request by dgrammatiko**
Pull Request for Issue https://github.com/joomla/joomla-cms/issues/19427

### Summary of Changes
- Moving the olde code to ES6 and custom elements
- Remove chosen (=== dead code)

### Testing Instructions

Go to admin->modules, edit a module, change the position and check the order field

### CODE REVIEW

File: build/webcomponents/js/field-module-order/field-module-order.js [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+			selectNode.appendChild(node);
+			selectNode.parentNode.innerHtml = '';
+			selectNode.parentNode.appendChild(selectNode);
+
+			i++;
+		}
+	}
+
+	getNewOrder (originalPos) {
+		console.log('dfgd')
```
Comment (anuragteapot) [inline build/webcomponents/js/field-module-order/field-module-order.js]: I think this one is debugging code. No need.

**Comment by anuragteapot** [comment]
@dgt41 what, if none of the options are selected its shows like 

screenshot from 2018-02-13 21-06-26

And this one is the previous image without the custom element.

screenshot from 2018-02-13 21-06-08

Also, is search option removed?

File: media/system/webcomponents/js/joomla-field-module-order-es5.js [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+				selectNode.parentNode.innerHtml = '';
+				selectNode.parentNode.appendChild(selectNode);
+
+				i++;
+			}
+		}
+	}, {
+		key: 'getNewOrder',
+		value: function getNewOrder(originalPos) {
+			console.log('dfgd');
```
Comment (anuragteapot) [inline media/system/webcomponents/js/joomla-field-module-order-es5.js]: Here also

File: media/system/webcomponents/js/joomla-field-module-order.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+			selectNode.appendChild(node);
+			selectNode.parentNode.innerHtml = '';
+			selectNode.parentNode.appendChild(selectNode);
+
+			i++;
+		}
+	}
+
+	getNewOrder (originalPos) {
+		console.log('dfgd')
```
Comment (anuragteapot) [inline media/system/webcomponents/js/joomla-field-module-order.js]: Here also

**Comment by dgrammatiko** [comment]
> Also, is search option removed?

Yes

File: build/webcomponents/js/field-module-order/field-module-order.js (hunk 1 above)
Comment (anuragteapot) [reply build/webcomponents/js/field-module-order/field-module-order.js]: Please remove <code>console.log('dfgd')</code>

Comment (dgrammatiko) [reply build/webcomponents/js/field-module-order/field-module-order.js]: removed

File: build/webcomponents/js/field-module-order/field-module-order.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+		let i = 0;
+		let x;
+		let item;
+
+		for ( x in source ) {
+			if (!source.hasOwnProperty(x)) { continue; }
+
+			item = source[ x ];
+
+			node = document.createElement('option');
```
Comment (anuragteapot) [inline build/webcomponents/js/field-module-order/field-module-order.js]: @dgt41 <code>node is not defined</code> error in console.

File: media/system/webcomponents/js/joomla-field-module-order.js [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+		let i = 0;
+		let x;
+		let item;
+
+		for ( x in source ) {
+			if (!source.hasOwnProperty(x)) { continue; }
+
+			item = source[ x ];
+
+			node = document.createElement('option');
```
Comment (anuragteapot) [inline media/system/webcomponents/js/joomla-field-module-order.js]: node not defined

**Comment by anuragteapot** [comment]
@dgt41 Please check my comments here and correct this :)

**Comment by dgrammatiko** [comment]
@Anu1601CS good catch!

**Comment by anuragteapot** [comment]
Thanks

**Comment by anuragteapot** [comment]
@dgt41 All things are good.
But, is this correct?

screenshot from 2018-02-16 03-43-16

<code>data-attr="class=" module-ajax-ordering""</code>

**Comment by Anu1601CS** [comment]
I have tested this item :white_check_mark: successfully on 6a625e8d23636304512a5db069608b38e7997d89<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19644">issues.joomla.org/tracker/joomla-cms/19644</a>.</sub>

**Comment by anuragteapot** [comment]
I have tested this item :white_check_mark: successfully on 2617b00e5e4f1e8ae4066cf325a1ca53786d4c6d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19644">issues.joomla.org/tracker/joomla-cms/19644</a>.</sub>

**Comment by anuragteapot** [comment]
@astridx Please test this.

**Comment by astridx** [comment]
Maybe I do not understand this PR right? The module field always shows only one entry, that is <tt>None</tt>. I have got three modules published on the position?

And the position field did not contain the text <tt>type or select a position</tt>

modules articles most read test administration

**Comment by dgrammatiko** [comment]
@astridx what are you getting in the main repo (i mean without the code of this PR)?

**Comment by dgrammatiko** [comment]
> And the position field did not contain the text type or select a position

Yes that was functionality from chosen.js, we will eventually replicate it in the custom element, but right now it's not there yet

**Comment by anuragteapot** [comment]
@dgt41 what for Ordering it always shows <code>none</code> ?

Without PR it shows 
screenshot from 2018-03-01 23-03-20

**Comment by dgrammatiko** [comment]
Ok, there's a bug here, will push some code later on

**Comment by dgrammatiko** [comment]
@Anu1601CS @astridx it should be ok now

File: media/system/webcomponents/js/joomla-field-module-order-es5.js [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+		value: function getNewOrder(originalPosition) {
+			var url = this.getAttribute('data-url');
+			var clientId = this.getAttribute('data-client-id');
+			var originalOrder = this.getAttribute('data-ordering');
+			var name = this.getAttribute('data-name');
+			var attr = this.getAttribute('data-client-attr') ? this.getAttribute('data-client-attr') : 'custom-select';
+			var id = this.getAttribute('id') + '_1';
+			var orders = [];
+			var that = this;
+			console.log(originalPosition);
```
Comment (anuragteapot) [inline media/system/webcomponents/js/joomla-field-module-order-es5.js]: @dgt41 please remove debug code.

File: media/system/webcomponents/js/joomla-field-module-order-es5.js [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+					if (response) {
+						response = JSON.parse(response);
+
+						/** Check if everything is OK **/
+						if (response.data.length > 0) {
+							for (var i = 0; i < response.data.length; ++i) {
+								orders[i] = response.data[i].split(',');
+							}
+
+							console.log(orders);
```
Comment (anuragteapot) [inline media/system/webcomponents/js/joomla-field-module-order-es5.js]: remove debug code.

File: media/system/webcomponents/js/joomla-field-module-order.js [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+	getNewOrder (originalPosition) {
+		const url = this.getAttribute('data-url');
+		const clientId = this.getAttribute('data-client-id');
+		const originalOrder = this.getAttribute('data-ordering');
+		const name = this.getAttribute('data-name');
+		const attr = this.getAttribute('data-client-attr') ? this.getAttribute('data-client-attr') : 'custom-select';
+		const id = this.getAttribute('id') + '_1';
+		const orders = [];
+		const that = this;
+console.log(originalPosition)
```
Comment (anuragteapot) [inline media/system/webcomponents/js/joomla-field-module-order.js]: remove debug code

File: media/system/webcomponents/js/joomla-field-module-order.js [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+						response = JSON.parse(response);
+
+						/** Check if everything is OK **/
+						if (response.data.length > 0) {
+							for (let i = 0; i < response.data.length; ++i) {
+								orders[i] = response.data[i].split(',');
+							}
+
+							console.log(orders)
+							that.writeDynaList({
```
Comment (anuragteapot) [inline media/system/webcomponents/js/joomla-field-module-order.js]: remove debug code.

**Comment by anuragteapot** [comment]
All these things seem OK form my side. :)

**Comment by dgrammatiko** [comment]
console.log removed

**Comment by anuragteapot** [comment]
@dgt41 Thanks

**Comment by anuragteapot** [comment]
I have tested this item :white_check_mark: successfully on 8031d2851d9c6a4ba906ff7a5665804aa10d3894<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19644">issues.joomla.org/tracker/joomla-cms/19644</a>.</sub>

**Comment by astridx** [comment]
I have tested this item :white_check_mark: successfully on 8031d2851d9c6a4ba906ff7a5665804aa10d3894<br /><br />When I assigned the branch, I first saw only None in the field module ordering. I then ran <tt>npm install</tt
modules articles most read test administration
> and executed all node build commands. Then it worked.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19644">issues.joomla.org/tracker/joomla-cms/19644</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19644">issues.joomla.org/tracker/joomla-cms/19644</a>.</sub>

**Comment by laoneo** [comment]
Can you fix conflicts here?

**Comment by dgrammatiko** [comment]
@laoneo done

File: layouts/joomla/form/field/moduleorder.php [hunk 10]
```diff
@@ -47,26 +47,35 @@
...
 
-// Initialize JavaScript field attributes.
-$attr .= !empty($onchange) ? ' onchange="' . $onchange . '"' : '';
+if ($onchange)
+{
+	$attributes['onchange'] = 'onchange="' . $onchange . '"';
+}
 
-HTMLHelper::_('script', 'system/fields/moduleorder.min.js', array('version' => 'auto', 'relative' => true));
+HTMLHelper::_('webcomponent', ['joomla-field-module-order' => 'system/webcomponents/joomla-field-module-order.min.js'], ['version' => 'auto', 'relative' => true]);
```
Comment (laoneo) [inline layouts/joomla/form/field/moduleorder.php]: This needs to be adopted?

Comment (anuragteapot) [reply layouts/joomla/form/field/moduleorder.php]: @dgt41 can you please fix this ?

**Comment by anuragteapot** [comment]
I have tested this item :white_check_mark: successfully on e2ac2c1b449eb36c9c8d20f5bb533afbb01d7828<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19644">issues.joomla.org/tracker/joomla-cms/19644</a>.</sub>

**Comment by ghost** [comment]
@astridx can you please retest?

**Pull request merged**